File: tella-locking-ui/src/main/java/com/hzontal/tella_locking_ui/ui/PasswordUnlockActivity.java
package com.hzontal.tella_locking_ui.ui;

import android.os.Bundle;
import android.widget.Button;

import androidx.appcompat.app.AppCompatActivity;
import com.hzontal.tella_locking_ui.R;

import org.hzontal.tella.keys.config.IUnlockRegistryHolder;
import org.hzontal.tella.keys.config.UnlockRegistry;
import org.hzontal.tella.keys.config.UnlockResult;

import javax.crypto.spec.PBEKeySpec;

public class PasswordUnlockActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.unlock_activity);

        Button unlockButton = findViewById(R.id.unlock_button);
        unlockButton.setText("Password");
        unlockButton.setOnClickListener(v -> handleUnlock("password"));

        Button badButton = findViewById(R.id.bad_button);
        badButton.setOnClickListener(v -> handleUnlock("xxx"));
    }

    private void handleUnlock(String secret) {
        IUnlockRegistryHolder holder = (IUnlockRegistryHolder) getApplicationContext();
        UnlockRegistry registry = holder.getUnlockRegistry();

        registry.putResult(PasswordUnlockActivity.class.getName(),
                new UnlockResult(new PBEKeySpec(secret.toCharArray())));

        setResult(RESULT_OK);
        finish();
    }
}

File: tella-locking-ui/src/main/java/com/hzontal/tella_locking_ui/ui/password/PasswordUnlockActivity.kt
package com.hzontal.tella_locking_ui.ui.password

import android.os.Bundle
import android.view.KeyEvent
import android.widget.ImageView
import androidx.core.content.ContextCompat
import androidx.core.view.isVisible
import com.hzontal.tella_locking_ui.R
import com.hzontal.tella_locking_ui.ReturnActivity
import com.hzontal.tella_locking_ui.TellaKeysUI
import com.hzontal.tella_locking_ui.common.extensions.onChange
import com.hzontal.tella_locking_ui.ui.password.base.BasePasswordActivity
import org.hzontal.tella.keys.MainKeyStore
import org.hzontal.tella.keys.key.MainKey
import timber.log.Timber
import javax.crypto.spec.PBEKeySpec

private const val TAG = "PasswordUnlockActivity"

class PasswordUnlockActivity : BasePasswordActivity() {


    private lateinit var backBtn: ImageView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initView()
    }

    private fun initView() {

        topImageView.background = ContextCompat.getDrawable(this, R.drawable.tella_logo_dark_bg)
        enterPasswordTextView.isVisible = false
        passwordLeftButton.isVisible = false
        passwordRightButton.isVisible = false
        passwordEditText.setOnKeyListener { _, keyCode, keyEvent ->
            if (keyEvent.action == KeyEvent.ACTION_DOWN && keyCode == KeyEvent.KEYCODE_ENTER) {
                if (isHiLighted)
                    onSuccessSetPassword(mPassword)
                return@setOnKeyListener true
            }
            return@setOnKeyListener false
        }
        when (returnActivity) {
            ReturnActivity.SETTINGS.getActivityOrder() -> {
                backBtn = findViewById(R.id.backBtn)
                backBtn.isVisible = true
                backBtn.setOnClickListener { finish() }
                passwordMsgTextView.text = getString(R.string.LockPasswordSet_Settings_EnterCurrentPassword)
                passwordEditText.hint = getString(R.string.LockPasswordSet_Settings_EnterCurrentPassword)
            }
            ReturnActivity.CAMOUFLAGE.getActivityOrder() -> {
                backBtn = findViewById(R.id.backBtn)
                backBtn.isVisible = true
                backBtn.setOnClickListener { finish() }
                passwordMsgTextView.text = getString(R.string.LockPasswordSet_Settings_EnterCurrentPasswordToChangeCamouflage)
                passwordEditText.hint = getString(R.string.LockPasswordSet_Settings_EnterCurrentPasswordToChangeCamouflage)
            }
            else -> {
                passwordMsgTextView.text = getText(R.string.UnlockPassword_Message_EnterPassword)
                passwordEditText.hint = getString(R.string.UnlockPassword_Message_EnterPassword)
                passwordEditText.onChange {
                    passwordMsgTextView.text = getText(R.string.UnlockPassword_Message_EnterPassword)
                }
            }
        }
    }

    override fun onSuccessSetPassword(password: String) {
        TellaKeysUI.getMainKeyStore().load(config.wrapper, PBEKeySpec(password.toCharArray()), object : MainKeyStore.IMainKeyLoadCallback {
            override fun onReady(mainKey: MainKey) {
                Timber.d("*** MainKeyStore.IMainKeyLoadCallback.onReady")
                TellaKeysUI.getMainKeyHolder().set(mainKey);
                onSuccessfulUnlock()
                finish()
            }

            override fun onError(throwable: Throwable) {
                Timber.d(throwable, "*** MainKeyStore.UnlockRegistry.IUnlocker.onError")
                onFailureSetPassword(getString(R.string.LockPasswordSet_Message_Error_IncorrectPassword))
                TellaKeysUI.getCredentialsCallback().onUnSuccessfulUnlock(TAG, throwable)
            }
        })
    }

    override fun onFailureSetPassword(error: String) {
        passwordMsgTextView.text = error
        passwordEditText.setTextColor(ContextCompat.getColor(this, R.color.wa_red_error));
    }
}